cranelift/codegen/src/souper_harvest.rs

```diff
@@ -0,0 +1,525 @@
+                    }
+                    (ir::Opcode::IshlImm, ir::InstructionData::BinaryImm64 { imm, .. }) => {
+                        let a = arg(allocs, 0);
+                        let value: i64 = (*imm).into();
+                        let value: i128 = value.into();
+                        let b = ast::Constant {
+                            value,
+                            r#type: souper_type_of(&func.dfg, val),
+                        }
+                        .into();
+                        ast::Instruction::Shl { a, b }.into()
+                    }
+                    (ir::Opcode::Sshr, _) => {
+                        let a = arg(allocs, 0);
+                        let b = arg(allocs, 1);
+                        ast::Instruction::Ashr { a, b }.into()
+                    }
+                    (ir::Opcode::SshrImm, ir::InstructionData::BinaryImm64 { imm, .. }) => {
+                        let a = arg(allocs, 0);
+                        let value: i64 = (*imm).into();
+                        let value: i128 = value.into();
+                        let b = ast::Constant {
+                            value,
+                            r#type: souper_type_of(&func.dfg, val),
+                        }
+                        .into();
+                        ast::Instruction::Ashr { a, b }.into()
+                    }
+                    (ir::Opcode::Ushr, _) => {
+                        let a = arg(allocs, 0);
+                        let b = arg(allocs, 1);
+                        ast::Instruction::Lshr { a, b }.into()
+                    }
+                    (ir::Opcode::UshrImm, ir::InstructionData::BinaryImm64 { imm, .. }) => {
+                        let a = arg(allocs, 0);
+                        let value: i64 = (*imm).into();
+                        let value: i128 = value.into();
+                        let b = ast::Constant {
+                            value,
+                            r#type: souper_type_of(&func.dfg, val),
+                        }
+                        .into();
+                        ast::Instruction::Lshr { a, b }.into()
+                    }
+                    (ir::Opcode::Select, _) => {
+                        let a = arg(allocs, 0);
+                        let b = arg(allocs, 1);
+                        let c = arg(allocs, 2);
+                        ast::Instruction::Select { a, b, c }.into()
+                    }
+                    (ir::Opcode::Uextend, _) => {
+                        let a = arg(allocs, 0);
+                        ast::Instruction::Zext { a }.into()
+                    }
+                    (ir::Opcode::Sextend, _) => {
+                        let a = arg(allocs, 0);
+                        ast::Instruction::Sext { a }.into()
+                    }
+                    (ir::Opcode::Trunc, _) => {
+                        let a = arg(allocs, 0);
+                        ast::Instruction::Trunc { a }.into()
```

Comment:
Me not knowing Souper well: how is the size to truncate to inferred?

Reply:
`truncate` instructions require a type annotation on the result. When harvesting, we always add the type annotation to every result (see the `souper_type_of` call after this `match`). We could elide most of them, allowing souper to infer them, but it seems like extra work and maintenance for very little benefit.